src/crypto/index.js

```diff
@@ -965,51 +966,62 @@ Crypto.prototype.importRoomKeys = function(keys) {
     );
 };
 
-Crypto.prototype._backupPayloadForSession = function(
-    senderKey, forwardingCurve25519KeyChain,
-    sessionId, sessionKey, keysClaimed,
-    exportFormat,
-) {
-    // new session.
-    const session = new Olm.InboundGroupSession();
-    try {
-        if (exportFormat) {
-            session.import_session(sessionKey);
-        } else {
-            session.create(sessionKey);
-        }
-        if (sessionId != session.session_id()) {
-            throw new Error(
-                "Mismatched group session ID from senderKey: " +
-                    senderKey,
-            );
-        }
+Crypto.prototype._maybeSendKeyBackup = async function() {
+    if (!this._sendingBackups) {
+        this._sendingBackups = true;
+        while (1) {
+            if (!this.backupKey) {
+                this._sendingBackups = false;
+                return;
+            }
+            // FIXME: figure out what limit is reasonable
+            const sessions = await this._cryptoStore.getSessionsNeedingBackup(10);
+            if (!sessions.length) {
+                this._sendingBackups = false;
+                return;
+            }
+            const data = {};
+            for (const session of sessions) {
+                const room_id = session.sessionData.room_id;
+                if (data[room_id] === undefined)
+                    data[room_id] = {sessions: {}};
+
+                const sessionData = await this._olmDevice.exportInboundGroupSession(session.senderKey, session.sessionId, session.sessionData);
+                sessionData.algorithm = olmlib.MEGOLM_ALGORITHM;
+                delete sessionData.session_id;
+                delete sessionData.room_id;
+                const encrypted = this.backupKey.encrypt(JSON.stringify(sessionData));
+
+                data[room_id]['sessions'][session.sessionId] = {
+                    first_message_index: 1, // FIXME
+                    forwarded_count: (sessionData.forwardingCurve25519KeyChain || []).length,
+                    is_verified: false, // FIXME: how do we determine this?
+                    session_data: encrypted,
+                };
+            }
 
-        if (!exportFormat) {
-            sessionKey = session.export_session();
+            let successful = false;
+            do {
+                if (!this.backupKey) {
+                    this._sendingBackups = false;
+                    return;
+                }
+                try {
+                    await this._baseApis.sendKeyBackup(undefined, undefined, this.backupInfo.version, {rooms: data});
+                    successful = true;
+                    await this._cryptoStore.unmarkSessionsNeedingBackup(sessions);
+                }
+                catch (e) {
+                    console.log("send failed", e);
+                    // FIXME: pause
+                }
+            } while (!successful);
+            // FIXME: pause between iterations?
```

Comment:
to consolidate my comments from the individual commits, we should put the pause in, or just stop & retry next time a backup is triggered which would be my inclination.